### ISSUE TITLE
Need Documentation on how dex works.

### ISSUE THREAD

**Issue by meenakshi123** [dexidp/dex#1737]
I want to use dex where i need to authenticate user with IAM and my application is not using the standard oauth2 flow.I don't have client id and secret.i need to call a url which returns access token and id token and then i need to redirect that to another page.Is there something like this which i can go through..

**Comment by zicklag** [comment]
I'm needing to do something very similar or maybe the same as you and I just did some research on how that works. I'll try to reply with an easy to understand/short version, but the full spec for OpenID and how to get tokens from dex is [here](https://openid.net/specs/openid-connect-core-1_0.html).

Dex is an OpenID Connect provider, and the [implicity flow](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowSteps) model specifically is perfect for getting tokens from an SPA for example.

**Comment by meenakshi123** [comment]
Thanks zicklag for the reply.
I already have a url which gives me access token and id token but i am not sure how to use that in dex authentication.
Can you please provide more details on it. As it is quite urgent for me.

**Comment by zicklag** [comment]
I'm not sure I fully understand your setup.

Do you:

1. Have Dex setup so that it uses some external service to authenticate users such as GitHub **and** you want to allow users to log into your app using the Dex login form, or
2. You want to have Dex use *your application* as an authentication backend?

**Comment by meenakshi123** [comment]
I want to use dex for authentication.But i am not sure how to use it in my case..in my case i have only a url which gives me access token and id token..using id token i can get user identity but it does not redirect me automatically to application dashboard page..I have done same thing with ldap..which was fine..as we have ldap connector in dex.
But not sure about IAM as we don't have any connector.May be we need to customize it.

**Comment by zicklag** [comment]
What is IAM?

**Comment by meenakshi123** [comment]
identity and access mgmt

**Comment by meenakshi123** [comment]
may i know what you are trying to achieve.May be i can try to relate.

**Comment by zicklag** [comment]
I've just finished setting up Dex to point to an LDAP auth backend and successfully gotten the `id_token` from my application. The workflow is like this

1. Use hits my application ( they are unauthorized )
2. My app uses [node-openid-client](https://github.com/panva/node-openid-client) to discover the redirect URL for dex:
```js
let issuer = await Issuer.discover(
    "http://123.456.789.100:5556/.well-known/openid-configuration"
  );

  const client = new issuer.Client({
    client_id: "testapp",
    client_secret: "password",
    // This redirect URI must point to the application URL that you want Dex to redirect to after
    // the user is authenticated.
    redirect_uris: ["http://123.456.789.100:3002/callback"],
    response_types: ["id_token"]
  });

  // My app will redirect to this URL to get the user authenticated
  const url = client.authorizationUrl({
    scope: "openid groups profile email",
    nonce: "hellononce",
    response_mode: "query",
    display: "popup"
  });
```
3. My app redirect's the user to Dex
4. Dex authenticates the user and redirects back to my app using the URL in the `redirect_uris` section in the above code sample. Dex will redirect to that URL with the tokens in the hastag: `http://123.456.789.100:3002/callback#id_token=[token]&access_token=[other_token]`.
5. My SPA app reads the tokens out of the hashtag and uses the `id_token` to make requests to my backend API

Is this similar to what you are trying to accomplish?

**Comment by meenakshi123** [comment]
in my case i don't have client id and secret. will it work if i don't pass it?or can it be some random text?

**Comment by zicklag** [comment]
Ah, you set the client id and secret in the dex config yaml in the `staticClients` section:

```yaml
staticClients:
  - id: my-client-id
    redirectURIs:
      - http://123.456.789.100:3002/callback
    name: My Client ID
    secret: my-client-secret
```

**Comment by meenakshi123** [comment]
ok..I will try this..And is your code similar to other connectors?Do you have Open,LoginURL,HAndleCallback this method?

**Comment by zicklag** [comment]
I don't know what you mean exactly...

I haven't seen any other connector's code yet. In my use-case everything happens in the JavaScript frontend ( or it will once I've got it all put together ).

The JavaScript frontend will realize check a cookie for the id_token, and if it doesn't find it it will redirect use the above code to calculate the Dex redirect URL, then it will set the web-page's `href` to that URL so Dex can authenticate the user. When dex redirect's back to the app, the JavaScript frontend will read the id_token from the hashtag and use Axios to make API requests with the token in the `Authorization` header.

**Comment by meenakshi123** [comment]
I see..i think the node-openid-client  approach is little bit different than the dex connector flow.If you want we can connect on hangout and can explain you better, if you don't mind.

**Comment by zicklag** [comment]
Just Sorry, because of network restrictions on my side hangouts don't work for me. Unfortunately this is the best place to discuss that I have access to.

**Comment by zicklag** [comment]
Not sure if this helps, but here is a sequence diagram:

![Untitled Diagram](https://user-images.githubusercontent.com/25393315/85782621-4f77a680-b6ec-11ea-9def-379c937656d0.png)

**Comment by zicklag** [comment]
If you have a specific step you have questions about then I can try to explain it better.

Also, are you doing this from a single paged application ( SPA )? Or are you doing it from something like a PHP web application or something similar.  I might be able to publish an example repo that shows how I'm doing it or how you could do it depending on what application stack you are using.

**Comment by meenakshi123** [comment]
No Issues..I am interested in step 4 and step 5.

**Comment by zicklag** [comment]
OK, so those are completely handled by Dex. After step 2, the user is presented with the Dex login screen:

![image](https://user-images.githubusercontent.com/25393315/85783627-1be94c00-b6ed-11ea-9b9f-b92c102143f2.png)

In my case I select LDAP and then I login with my username and password there:

![image](https://user-images.githubusercontent.com/25393315/85783718-30c5df80-b6ed-11ea-9c80-17db6937f05a.png)

---

Have you been able to get to that point so far?

**Comment by meenakshi123** [comment]
yes

**Comment by zicklag** [comment]
OK, so what happens when you submit the login form?

It should redirect back to your application, with the `id_token` ( and other stuff ) in the URL.

**Comment by meenakshi123** [comment]
This is my 1st request- Request URL: http://127.0.0.1:5556/dex/auth/idam?req=bfql54ekleejpfhpjk4uqxfb4
2nd request: http://127.0.0.1:5556/dex/auth/?client_id=&response_type=code&state=bfql54ekleejpfhpjk4uqxfb4
I think it is taking from example-app.In my dex-config.yaml, i have provided this:-oauth2:
  response_types: ["id_token"]

but i want to know where should i add my api call which returns me accestoken.

**Comment by zicklag** [comment]
To clarify, you don't need to use [node-openid-cient](https://github.com/panva/node-openid-client) at all. That just automates putting together the first request and helping with the workflow a bit, but it is easier to learn about how it works without involving it.

**Comment by meenakshi123** [comment]
This is fine.but in my case, we have one extra thing- to call a api and i am getting tokens from there and because of that i am confused. and even i don't have username password too.I h.ave a singal button when i click on that, it should call my api in that api it checks the computer certificated and based on that it returns id token and access token.Now the point is, using that token , dex authentication is not happening.

**Comment by zicklag** [comment]
So you have an API that will figure out whether or not a user is authenticated, and you want Dex to use that API to authenticate users? For example:

- I'm a user that goes to your app
- I click the "Login Button"
- Your app redirects me to Dex where there is just a "Login with [Your API name]"
- I click the button and, without even putting in a password that API authenticates me
- Dex redirects to the app that I went to initially

In that case you would have to write you own connector for Dex, which would require writing some Go and getting it merged into Dex. Or you could fork Dex for your own use and add the code in your fork. I don't know what it takes to implement a Dex connector, but you can see the list of built-in connectors [here](https://github.com/dexidp/dex/tree/master/connector).

**Comment by meenakshi123** [comment]
Right..I have created a new connector but i am little bit confused about the connector methods(Open,LoginURL, Callback).I am not sure how this works and what each method will do in my case.and these methods are required.

**Comment by zicklag** [comment]
Ah, OK, that makes more sense. I thought I was misunderstanding something.

When it comes to that, I unfortunately don't know much. :confused: I haven't implemented my own connector in Dex yet, nor have I written much Go.

**Comment by meenakshi123** [comment]
same case with me..If you find out any documentation about these methods or any blogs which are related to this.Please ping me.Thanks for your time and help.

**Comment by zicklag** [comment]
:+1: No problem. :)
